src/v3/core/address.test.js

```diff
@@ -58,5 +58,142 @@ describe("core/address", () => {
+        }).toThrow("widget: expected FooAddress, got: null");
+      });
+      it("rejects an address from a known module", () => {
+        const bar = BarAddress.fromParts(["hello"]);
+        expect(() => {
+          FooAddress.assertValid(bar, "widget");
+        }).toThrow("widget: expected FooAddress, got BarAddress:");
+      });
+      it("rejects an address from an unknown module", () => {
+        const wat = WatAddress.fromParts(["hello"]);
+        expect(() => {
+          FooAddress.assertValid(wat, "widget");
+        }).toThrow("widget: expected FooAddress, got:");
+      });
+      it("rejects a string that starts with the nonce but no separator", () => {
+        expect(() => {
+          FooAddress.assertValid("F/things\0", "widget");
+        }).toThrow("widget: expected FooAddress, got:");
+      });
+      it("rejects a string that does not end with a separator", () => {
+        const address = FooAddress.fromParts(["hello"]);
+        const truncated = address.substring(0, address.length - 2);
+        expect(() => {
+          FooAddress.assertValid(truncated, "widget");
+        }).toThrow("widget: expected FooAddress, got:");
+      });
+      it("rejects junk", () => {
+        expect(() => {
+          FooAddress.assertValid("junk", "widget");
+        }).toThrow("widget: expected FooAddress, got:");
+      });
+    });
+
+    describe("assertValidParts", () => {
+      const {FooAddress} = makeModules();
+      it("rejects `undefined`", () => {
+        expect(() => {
+          // $ExpectFlowError
+          FooAddress.assertValidParts(undefined, "widget");
+        }).toThrow("widget: expected array of parts, got: undefined");
+      });
+      it("rejects `null`", () => {
+        expect(() => {
+          // $ExpectFlowError
+          FooAddress.assertValidParts(null, "widget");
+        }).toThrow("widget: expected array of parts, got: null");
+      });
+      it("rejects an array containing `undefined`", () => {
+        expect(() => {
+          // $ExpectFlowError
+          FooAddress.assertValidParts(["hello", undefined, "world"], "widget");
+        }).toThrow(
+          "widget: expected array of parts, got undefined in: " +
+            '["hello",undefined,"world"]'
+        );
+      });
+      it("rejects an array containing `null`", () => {
+        expect(() => {
+          // $ExpectFlowError
+          FooAddress.assertValidParts(["hello", null, "world"], "widget");
+        }).toThrow(
+          "widget: expected array of parts, got null in: " +
+            '["hello",null,"world"]'
+        );
+      });
+      it("rejects an array with a string containing a NUL character", () => {
+        expect(() => {
+          FooAddress.assertValidParts(["hello", "n\0o", "world"], "widget");
+        }).toThrow(
+          "widget: part contains NUL character: " +
+            '"n\\u0000o" in ["hello","n\\u0000o","world"]'
+        );
+      });
+    });
+
+    describe("fromParts", () => {
```

Comment:
I note that a `fromParts` implementation that neglected to call `assertValidParts` would pass tests. please fix

Reply:
Good point; will fix.